internal/http/services/appprovider/appprovider.go

```diff
@@ -0,0 +1,232 @@
+// or submit itself to any jurisdiction.
+
+package ocmd
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+	"unicode/utf8"
+
+	gateway "github.com/cs3org/go-cs3apis/cs3/gateway/v1beta1"
+	rpc "github.com/cs3org/go-cs3apis/cs3/rpc/v1beta1"
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	"github.com/cs3org/reva/internal/http/services/ocmd"
+	"github.com/cs3org/reva/pkg/rgrpc/status"
+	"github.com/cs3org/reva/pkg/rgrpc/todo/pool"
+	"github.com/cs3org/reva/pkg/rhttp/global"
+	"github.com/cs3org/reva/pkg/sharedconf"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+)
+
+func init() {
+	global.Register("appprovider", New)
+}
+
+// Config holds the config options that need to be passed down to all ocdav handlers
+type Config struct {
+	Prefix         string `mapstructure:"prefix"`
+	GatewaySvc     string `mapstructure:"gatewaysvc"`
+	AccessTokenTTL int    `mapstructure:"access_token_ttl"`
+}
+
+func (c *Config) init() {
+	if c.Prefix == "" {
+		c.Prefix = "api/v0/wopi/open"
+	}
+	if c.AccessTokenTTL == 0 {
+		c.AccessTokenTTL = 86400
+	}
+	c.GatewaySvc = sharedconf.GetGatewaySVC(c.GatewaySvc)
+}
+
+type svc struct {
+	conf *Config
+}
+
+// New returns a new ocmd object
+func New(m map[string]interface{}, log *zerolog.Logger) (global.Service, error) {
+
+	conf := &Config{}
+	if err := mapstructure.Decode(m, conf); err != nil {
+		return nil, err
+	}
+	conf.init()
+
+	s := &svc{
+		conf: conf,
+	}
+	return s, nil
+}
+
+// Close performs cleanup.
+func (s *svc) Close() error {
+	return nil
+}
+
+func (s *svc) Prefix() string {
+	return s.conf.Prefix
+}
+
+func (s *svc) Unprotected() []string {
+	return []string{}
+}
+
+func (s *svc) Handler() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != http.MethodGet {
+			ocmd.WriteError(w, r, ocmd.APIErrorUnimplemented, "only GET requests are supported", errors.New("only GET requests are supported"))
+			return
+		}
+
+		s.handleWopiOpen(w, r)
+	})
+}
+
+// WopiResponse holds the various fields to be returned for a wopi open call
+type WopiResponse struct {
+	WopiClientURL  string `json:"wopiclienturl"`
+	AccessToken    string `json:"accesstoken"`
+	AccessTokenTTL int64  `json:"accesstokenttl"`
+}
+
+func (s *svc) handleWopiOpen(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	client, err := pool.GetGatewayServiceClient(s.conf.GatewaySvc)
+	if err != nil {
+		ocmd.WriteError(w, r, ocmd.APIErrorServerError, "error getting grpc gateway client", err)
+		return
+	}
+
+	info, errCode, err := s.getStatInfo(ctx, r.URL.Query().Get("fileId"), client)
+	if err != nil {
+		ocmd.WriteError(w, r, errCode, "error statting file", err)
+	}
+
+	openReq := gateway.OpenInAppRequest{
+		Ref:      &provider.Reference{ResourceId: info.Id},
+		ViewMode: getViewMode(info),
```

Comment:
From my viewpoint it would be desirable if somebody could also decide to open a file in readonly mode even if he has permissions to open it in readwrite mode.  (We didn't takle this in oCIS WOPI extension because we have only one open action in the Web UI until now, but this will be more flexible in the future https://github.com/owncloud/web/issues/5135)

Reply:
That's a very valid point, and the `viewMode` should indeed come from the client (the web UI) exactly like today it is an external parameter passed to the GRPC `open-in-app` command.

Reply:
The current logic could be used as a fallback if no `viewMode` was given. I originally used the logic to determine maximum permissions of a user with the stat call because we have no possibility to determine the ViewMode in ownCloud Web currently.

Reply:
Yep, that would be ideal. For now, the idea was to mimic the logic of ocis-wopiserver so we could use it directly with web. We also need to discuss changing the endpoint

Reply:
I also thought about how we could make this more generic, so that ownCloud Web just displays the available apps. Ideally this would be done by a extension (which is shipped with oCIS by default) or even integrated in Web directly. We probably should have a discussion with @kulmann and @dragotin about that topic. But I don't know if it is within the scope of this PR!?

Reply:
Yes that is incorporated in the PR! We had the `FindAppProviders` call which returns all apps for a specific mime type and now we also have a default app configured for each mime type. So we can expose the former to get a list of all apps and then the open call would have the app name as a parameter. Would be good to discuss this at the Friday meeting